Add support for selective exclusion and readmission of inputs

davidecavestro opened the pull request.

palazzem commented:
Thanks @davidecavestro for your contribution! I will go through the code to give you a feedback, but thanks for adding tests and documentation for the new API.

Just one question before I start the review: can you describe me here the goal of the new API and what are you trying to achieve? If you want, it's even fine if you write me a snippet about how you use the API. I'd like to keep the API smallest as possible, but I'm also happy to add everything that is needed.

It will help me to understand better the use case. Thank you very much!

davidecavestro commented:
I'm writing a Kivy app to control my home alarm.

I can share the whole project as FOSS (I'd do it anyway sooner or later).
It is basically a KivyMD KitchenSink demo stripped down and modified: I just need the time to do some cleanup and pushing it to github.

I'd like to achieve something similar to what you can do with alarm inputs from the web UI:
![image](https://user-images.githubusercontent.com/447917/83287705-63f67b00-a1e2-11ea-8442-85a46c2320af.png)

The kivy counterpart (still an early work in progress) at the moment is 
![image](https://user-images.githubusercontent.com/447917/83288399-93f24e00-a1e3-11ea-97f8-a2ee8947638f.png)

It reacts to a click on the checkbox calling the `exclude` command this way
  
```
    def exclude_sensor(self, sensor_id):
        def job():
            self.auth()

            pin = self.get_pin()
            with self.client.lock(pin) as c:
                try:
                    print('excluding sensor ', sensor_id)
                    c.exclude([sensor_id])
                finally:
                    self.client.unlock()
                vibrate(2)

        self.executor.submit(job)

```
Some notes:
* I just switched the UI from Italian to English and noticed they use "bypass" instead of "exclude"
* I'm still investigating on how the "bypass/exclusion" status is exposed by (in order to set the initial checkbox value)
* OFFTOPIC I've managed to explicitly call `unlock` since at the end of the `with` block  the lock is not automatically released

davidecavestro commented:
>I'm still investigating on how the bypass/exclusion status is exposed (in order to set the initial checkbox value)

Could it be useful adding something like raw query methods to expose the data as
returned by the server?

To this purpose this is an excerpt of the whole set of attributes exposed by the server api respectively for sectors and inputs.

_SECTORS_
```
'Active' = {bool} False
'ActivePartial' = {bool} False
'Max' = {bool} False
'Activable' = {bool} True
'ActivablePartial' = {bool} False
'InUse' = {bool} True
'Id' = {int} 7010
'Index' = {int} 0
'Element' = {int} 1
'CommandId' = {int} 0
'InProgress' = {bool} False
```
_INPUTS_
```
'Alarm' = {bool} False
'MemoryAlarm' = {bool} False
'Excluded' = {bool} False
'InUse' = {bool} True
'IsVideo' = {bool} False
'Id' = {int} 26114
'Index' = {int} 0
'Element' = {int} 1
'CommandId' = {int} 0
'InProgress' = {bool} False
```

palazzem commented:
Side note: I've also reconfigured CircleCI. It should run tests from your fork now.

davidecavestro commented on elmo/api/client.py:
```diff
@@ -214,6 +214,81 @@ def disarm(self, sectors=None):
+
+    @require_session
+    @require_lock
+    def admit(self, inputs):
```
I agree.

Indeed it was the first term that came into my mind.
Then I switched to _admit_ just because _include_ makes me ask: _what's the set an input is included into?_
The answer probably is: _the area which they belongs to_.
So they are not removed from it, but simply deactivated.

But since I was not happy with the _admit_ term, nor with _unexclude_, I'm going for _include_.

palazzem commented:
@davidecavestro checking this PR again. I'll take care of the merge so no action items on you!

palazzem commented:
@davidecavestro can you allow me to edit and push changes to your fork? It should be something along these lines: https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

davidecavestro commented:
Changes from mantainers seems already enabled, I've unchecked and rechecked
![image](https://user-images.githubusercontent.com/447917/129370540-4e316970-3f2a-4a19-a2fc-9fd567f81271.png)
I also invited you as a collaborator on my fork, just in case...

palazzem commented:
Weird, it tells me:
```
remote: Permission to davidecavestro/econnect-python.git denied to palazzem.
fatal: unable to access 'https://github.com/davidecavestro/econnect-python.git/': The requested URL returned error: 403
```

Can you try to give me admin access to your fork? So I can override any settings and update your master. Thanks!

davidecavestro commented:
>
> Weird, it tells me:
>
> remote: Permission to davidecavestro/econnect-python.git denied to palazzem.
> fatal: unable to access 'https://github.com/davidecavestro/econnect-python.git/': The requested URL returned error: 403
>
> I still see the invitation as pending
[image: image.png]

Can you try to give me admin access to your fork? So I can override any
> settings and update your master

I guess GH doesn't let me give administrator access from a personal
namespace (not an organisation)
I could try to transfer the fork to you (no prob for me, it was meant just
to propose patches to you)
[image: image.png]

Or if you prefer I can somewhat extract a patch from git.


Il giorno ven 13 ago 2021 alle ore 16:34 Emanuele Palazzetti <
***@***.***> ha scritto:

> Weird, it tells me:
>
> remote: Permission to davidecavestro/econnect-python.git denied to palazzem.
> fatal: unable to access 'https://github.com/davidecavestro/econnect-python.git/': The requested URL returned error: 403
>
> Can you try to give me admin access to your fork? So I can override any
> settings and update your master. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/palazzem/econnect-python/pull/55#issuecomment-898500591>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADNLLJC4AQ44ECEUOKYJWDT4UUNTANCNFSM4NNM5XCQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

palazzem commented:
@davidecavestro here we go. Permissions weren't updated for an access token I was using, all good now!

palazzem commented:
@davidecavestro contribution merged! thank you very much for pushing toward this direction. I'm moving forward to integrate the change we discussed. Thanks again!
